src/devices/sound/flt_biquad.h

```diff
@@ -0,0 +1,256 @@
+//  TYPE DEFINITIONS
+//**************************************************************************
+
+// ======================> filter_biquad_device
+
+class filter_biquad_device : public device_t, public device_sound_interface
+{
+public:
+	enum
+	{
+		LOWPASS1P = 0,
+		HIGHPASS1P,
+		LOWPASS,
+		HIGHPASS,
+		BANDPASS,
+		NOTCH,
+		PEAK,
+		LOWSHELF,
+		HIGHSHELF
+	};
+
+	filter_biquad_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock = 0);
+
+	// configuration
+	filter_biquad_device &setup(int type, double fc, double q, double gain)
+	{
+		m_type = type;
+		m_fc = fc;
+		m_q = q;
+		m_gain = gain;
+		return *this;
+	}
+
+	filter_biquad_device &filter_biquad_setup(int type, double fc, double q, double gain)
+	{
+		m_stream->update();
+		setup(type, fc, q, gain);
+		recalc();
+		return *this;
+	}
+
+	/* Setup a biquad filter structure based on an op-amp sallen-key lowpass circuit.
+	*
+	*                   .----------------------------.
+	*                   |                            |
+	*                  ---  c2                       |
+	*                  ---                           |
+	*                   |                            |
+	*            r1     |   r2                |\     |
+	*   In >----ZZZZ----+--ZZZZ---+--------+  | \    |
+	*                             |        '--|+ \   |
+	*                            ---  c1      |   >--+------> out
+	*                            ---       .--|- /   |
+	*                             |        |  | /    |
+	*                            gnd       |  |/     |
+	*                                      |         |
+	*                                      |   r4    |
+	*                                      +--ZZZZ---'
+	*                                      |
+	*                                      Z
+	*                                      Z r3
+	*                                      Z
+	*                                      |
+	*                                     gnd
+	*/
+	filter_biquad_device &opamp_sk_lowpass_setup(double r1, double r2, double r3, double r4, double c1, double c2)
+	{
+		if ((r1 == 0) || (r2 == 0) || (r3 == 0) || (r4 == 0) || (c1 == 0) || (c2 == 0))
+		{
+			fatalerror("filter_biquad_device::opamp_sk_lowpass_setup() - no parameters can be 0; parameters were: r1: %f, r2: %f, r3: %f, r4: %f, c1: %f, c2: %f", r1, r2, r3, r4, c1, c2); /* Filter can not be setup.  Undefined results. */
+		}
+
+		// note: if R3 doesn't exist, pass a value of RES_M(999.99) or the like, i.e. an 'infinite resistor'
+		double const gain = (r3 + r4) / r3;
+		double const fc = 1.0 / (2 * M_PI * sqrt(r1 * r2 * c1 * c2));
+		double const q = sqrt(r1 * r2 * c1 * c2) / ((r1 * c1) + (r2 * c1) + ((r2 * c2) * (1.0 - gain)));
+#ifdef FLT_BIQUAD_DEBUG_SETUP
+		fprintf(stderr,"filter_biquad_device::opamp_sk_lowpass_setup() yields: fc = %f, Q = %f, gain = %f\n", fc, q, gain); fflush(stderr);
```

Comment:
If this could be called before `logerror` works, use `osd_printf_verbose` or `osd_printf_debug`.